lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -2018,6 +2019,25 @@ ParseResult FIRStmtParser::parseNode() {
       parseOptionalInfo(info, subOps))
     return failure();
 
+  // Error out if the node type, with an optional outer flip stripped,
+  // is not passive. Note: this is less restrictive than the FIRRTL
+  // spec to accomodate for situations where the node is something
+  // weird like a module output or an instance input.
+  auto initializerType = initializer.getType().cast<FIRRTLType>();
+  auto underlyingType = initializerType;
+  auto flip = initializerType.dyn_cast_or_null<FlipType>();
+  if (flip)
+    underlyingType = flip.getElementType();
+  if (!underlyingType.isPassive()) {
```

Comment:
Because of the canonicalization rules for types (whereby the body of a flip is always passive), I think this is equivalent to:

`if (!initializerType.isPassiveType() && !initializerType.isa<FlipType>())`

That said, are nodes allowed on bundles with analog types?  If so, that will cause its own problem.  I wonder if this check is worth it for us: it would be straight-forward to just always call `convertToPassive` without the check, just like we do for primitives.

I'm ok with whatever you resolve here.

Reply:
@clattner wrote:
> That said, are nodes allowed on bundles with analog types?

I checked... It looks like this is a weird edge case in the SFC:

1. Nodes cannot have analog type (at its outermost type)
2. Nodes must be passive (analog is passive as it's defined in CIRCT, too)
3. Attach can only involve ports, instance ports, or wires (this is the SFC's "kind" system rearing its head)

The combination of these three results in the inability to create a node that is analog, but the ability to create an aggregate node that contains analog. However, the kind checks disable any use of an analog in the node aggregate.

@jackkoenig: do you have any thoughts on why checks (1) and (3) exist?

I think a reasonable solution for us is to continue to do what we're doing: make nodes respect (2) and ignore (1) and (3). We'll accept more than the SFC, but I think that's okay. I think this is already what we're doing anyway. Alternatively, it's pretty easy to add the check of (1) if we want it. (3) is way trickier and should be dealt with during parsing if at all. (Leaning towards ignoring this.)

@clattner wrote:
> ... it would be straight-forward to just always call convertToPassive without the check, just like we do for primitives.
That would work, too. I'd prefer to accept more circuits than the SFC, but just minimally more. I'll come up with some solution. (Leaning towards leaving the check in).

Reply:
And, good catch about how to simplify the logic based on type canonicalization. :+1: Using this now.

I've updated the commit to add a check of (1) during parsing. (1) is allowed once you're inside CIRCT, however. (3) is never checked.

Reply:
The fact that you can have `Analog` as part of an aggregate node sounds like an oversight. The basic reasoning (which may not be relevant in CIRCT) is that there is directionality in node definitions. The flow is one way from right-hand side to the node at definition, and then after that you cannot connect to or otherwise "drive" a node. Since `attach` is a bidirectional operation, it cannot operate on nodes, thus it also doesn't make sense to have `Analog` in nodes. I guess you could still disallow (3) while allowing Analog in nodes though.

Reply:
> I think a reasonable solution for us is to continue to do what we're doing: make nodes respect (2) and ignore (1) and (3). We'll accept more than the SFC, but I think that's okay. I think this is already what we're doing anyway. 

I completely agree with you.